Archive/Wiedijk100Theorems/BuffonsNeedle.lean

```diff
@@ -0,0 +1,390 @@
+-/
+lemma integral_min_eq_two_mul :
+    ∫ θ in (0)..π, min d (θ.sin * l) = 2 * ∫ θ in (0)..π / 2, min d (θ.sin * l) := by
+  rw [← intervalIntegral.integral_add_adjacent_intervals (b := π / 2) (c := π)]
+  conv => lhs; arg 2; arg 1; intro θ; rw [← neg_neg θ, Real.sin_neg]
+
+  simp_rw [intervalIntegral.integral_comp_neg fun θ => min d (-θ.sin * l), ← Real.sin_add_pi,
+    intervalIntegral.integral_comp_add_right (fun θ => min d (θ.sin * l)), add_left_neg,
+    (by ring : -(π / 2) + π = π / 2), two_mul]
+
+  all_goals exact intervalIntegrable_min_const_sin_mul d l _ _
+
+/--
+  The first of two adjacent integrals in the long case. In the range `(0)..(d / l).arcsin`, we
+  have that `θ.sin * l ≤ d`, and thus the integral is `∫ θ in (0)..(d / l).arcsin, θ.sin * l`.
+-/
+lemma integral_zero_to_arcsin_min :
+    ∫ θ in (0)..(d / l).arcsin, min d (θ.sin * l) = (1 - (1 - (d / l) ^ 2).sqrt) * l := by
+  have : Set.EqOn (fun θ => min d (θ.sin * l)) (Real.sin · * l) (Set.uIcc 0 (d / l).arcsin) := by
+    intro θ ⟨hθ₁, hθ₂⟩
+    have : 0 ≤ (d / l).arcsin := Real.arcsin_nonneg.mpr (div_nonneg hd.le hl.le)
+    simp only [sup_eq_max, inf_eq_min, min_eq_left this, max_eq_right this] at hθ₁ hθ₂
+
+    have hθ_mem : θ ∈ Set.Ioc (-(π / 2)) (π / 2) := by
+      exact ⟨lt_of_lt_of_le (neg_lt_zero.mpr (div_pos Real.pi_pos two_pos)) hθ₁,
+        le_trans hθ₂ (d / l).arcsin_mem_Icc.right⟩
+    simp_rw [min_eq_right ((le_div_iff hl).mp ((Real.le_arcsin_iff_sin_le' hθ_mem).mp hθ₂))]
+
+  rw [intervalIntegral.integral_congr this, intervalIntegral.integral_mul_const, integral_sin,
+    Real.cos_zero, Real.cos_arcsin]
+
+/--
+  The second of two adjacent integrals in the long case. In the range `(d / l).arcsin..(π / 2)`, we
+  have that `d ≤ θ.sin * l`, and thus the integral is `∫ θ in (d / l).arcsin..(π / 2), d`.
+-/
+lemma integral_arcsin_to_pi_div_two_min (h : d ≤ l) :
+    ∫ θ in (d / l).arcsin..(π / 2), min d (θ.sin * l) = (π / 2 - (d / l).arcsin) * d := by
+  have : Set.EqOn (fun θ => min d (θ.sin * l)) (fun _ => d) (Set.uIcc (d / l).arcsin (π / 2)) := by
+    intro θ ⟨hθ₁, hθ₂⟩
+
+    wlog hθ_ne_pi_div_two : θ ≠ π / 2
+    · simp only [ne_eq, not_not] at hθ_ne_pi_div_two
+      simp only [hθ_ne_pi_div_two, Real.sin_pi_div_two, one_mul, min_eq_left h]
+
+    simp only [sup_eq_max, inf_eq_min, min_eq_left (d / l).arcsin_le_pi_div_two,
+      max_eq_right (d / l).arcsin_le_pi_div_two] at hθ₁ hθ₂
+
+    have hθ_mem : θ ∈ Set.Ico (-(π / 2)) (π / 2) := by
+      exact ⟨le_trans (Real.arcsin_mem_Icc (d / l)).left hθ₁, lt_of_le_of_ne hθ₂ hθ_ne_pi_div_two⟩
+    simp_rw [min_eq_left ((div_le_iff hl).mp ((Real.arcsin_le_iff_le_sin' hθ_mem).mp hθ₁))]
+
+  rw [intervalIntegral.integral_congr this, intervalIntegral.integral_const, smul_eq_mul]
+
+/--
+  Buffon's Needle, the long case (`d ≤ l`).
+-/
+theorem buffon_long (h : d ≤ l) :
+    ℙ[N l B] = (2 * l) / (d * π) - 2 / (d * π) * ((l^2 - d^2).sqrt + d * (d / l).arcsin) + 1 := by
```

Comment:
Could you adjust the uses of `.sqrt` to use the new notation?

Reply:
Done, also merged in master.